### PR TITLE
Allow use array for labels in speech json

### PR DESCRIPTION
**Pull request by ZhilkinSerg**
#### Summary

```SUMMARY: Infrastructure "Allow use array for labels in speech json"```

#### Purpose of change

Allow use array for labels in speech json

#### Describe the solution

Updated speech json loading code to support both single string and array of strings.

#### Additional context

Prerequisite for #35648

### CODE REVIEW

File: src/speech.cpp [hunk 1]
```diff
@@ -13,18 +13,28 @@ static SpeechBubble nullSpeech = { no_translation( "INVALID SPEECH" ), 0 };
 
 void load_speech( JsonObject &jo )
 {
-    const std::string label = jo.get_string( "speaker" );
+    std::vector<std::string> labels;
+    if( jo.has_string( "speaker" ) ) {
```
Comment (BevapDin) [inline src/speech.cpp]: You could use `JsonObject::get_tags`, which will read a single string or an array of strings and return that as `std::set`. 

```C++
for( const std::string &label : jo.get_tags( "speaker" ) ) {
```
(Note as it returns a `std::set`, the order of the array elements may not be kept, if that's important.)

Comment (ZhilkinSerg) [reply src/speech.cpp]: Thanks a lot.

File: src/speech.cpp [hunk 2]
```diff
@@ -13,18 +13,17 @@ static SpeechBubble nullSpeech = { no_translation( "INVALID SPEECH" ), 0 };
...
     jo.read( "sound", sound );
     const int volume = jo.get_int( "volume" );
-
-    std::map<std::string, std::vector<SpeechBubble>>::iterator speech_type = speech.find( label );
-    // Construct a vector matching the label if needed.
-    if( speech_type == speech.end() ) {
-        speech_type = speech.emplace( label, std::vector<SpeechBubble>() ).first;
+    for( const std::string &label : jo.get_tags( "speaker" ) ) {
+        std::map<std::string, std::vector<SpeechBubble>>::iterator speech_type = speech.find( label );
+        // Construct a vector matching the label if needed.
```
Comment (BevapDin) [inline src/speech.cpp]: And this map lookup can be simplified via `std::map::operator[]` (it creates the value using the default constructor if the key does not exist within the map.)
```C++
speech[label].emplace_back( SpeechBubble{ sound, volume } );
```
And if you're using `emplace_back` (or `emplace`), that functions already constructs the object for you. You don't give it a full object, only its constructor arguments:
```C++
speech[label].emplace_back( sound, volume );
```

File: src/speech.cpp [hunk 3]
```diff
@@ -13,18 +13,17 @@ static SpeechBubble nullSpeech = { no_translation( "INVALID SPEECH" ), 0 };
...
     translation sound;
     jo.read( "sound", sound );
     const int volume = jo.get_int( "volume" );
-
-    std::map<std::string, std::vector<SpeechBubble>>::iterator speech_type = speech.find( label );
-    // Construct a vector matching the label if needed.
-    if( speech_type == speech.end() ) {
-        speech_type = speech.emplace( label, std::vector<SpeechBubble>() ).first;
+    for( const std::string &label : jo.get_tags( "speaker" ) ) {
+        std::map<std::string, std::vector<SpeechBubble>>::iterator speech_type = speech.find( label );
```
Comment (ZhilkinSerg) [inline src/speech.cpp]:
```suggestion
        speech[label].emplace_back( sound, volume );
```

File: src/speech.cpp (hunk 2 above)
Comment (ZhilkinSerg) [inline src/speech.cpp]:
```suggestion
```

File: src/speech.cpp [hunk 4]
```diff
@@ -13,18 +13,17 @@ static SpeechBubble nullSpeech = { no_translation( "INVALID SPEECH" ), 0 };
...
+        // Construct a vector matching the label if needed.
+        if( speech_type == speech.end() ) {
```
Comment (ZhilkinSerg) [inline src/speech.cpp]:
```suggestion
```

File: src/speech.cpp [hunk 5]
```diff
@@ -13,18 +13,17 @@ static SpeechBubble nullSpeech = { no_translation( "INVALID SPEECH" ), 0 };
...
+            speech_type = speech.emplace( label, std::vector<SpeechBubble>() ).first;
```
Comment (ZhilkinSerg) [inline src/speech.cpp]:
```suggestion
```

File: src/speech.cpp [hunk 6]
```diff
@@ -13,18 +13,17 @@ static SpeechBubble nullSpeech = { no_translation( "INVALID SPEECH" ), 0 };
...
+        }
```
Comment (ZhilkinSerg) [inline src/speech.cpp]:
```suggestion
```

File: src/speech.cpp [hunk 7]
```diff
@@ -13,18 +13,17 @@ static SpeechBubble nullSpeech = { no_translation( "INVALID SPEECH" ), 0 };
...
+        speech_type->second.emplace_back( SpeechBubble{ sound, volume } );
```
Comment (ZhilkinSerg) [inline src/speech.cpp]:
```suggestion
```

File: src/speech.cpp [hunk 8]
```diff
@@ -13,18 +13,12 @@ static SpeechBubble nullSpeech = { no_translation( "INVALID SPEECH" ), 0 };
...
     translation sound;
     jo.read( "sound", sound );
     const int volume = jo.get_int( "volume" );
-
-    std::map<std::string, std::vector<SpeechBubble>>::iterator speech_type = speech.find( label );
-    // Construct a vector matching the label if needed.
-    if( speech_type == speech.end() ) {
-        speech_type = speech.emplace( label, std::vector<SpeechBubble>() ).first;
+    for( const std::string &label : jo.get_tags( "speaker" ) ) {
+        speech[label].emplace_back( sound, volume );
```
Comment (kevingranade) [inline src/speech.cpp]: Compiler is failing to deduce how to invoke the constructor, this is what it did previously.
```suggestion
        speech[label].emplace_back( SpeechBubble{ sound, volume } );
```

**Pull request merged**